flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java

```diff
@@ -282,17 +282,15 @@ protected Configuration configuration() {
                         .build(),
                 CastTestSpecBuilder.testCastTo(BINARY(2))
                         .fromCase(BINARY(5), null, null)
-                        .fromCase(CHAR(3), "foo", new byte[] {102, 111, 111})
-                        .fromCase(VARCHAR(5), "Flink", new byte[] {70, 108, 105, 110, 107})
-                        // https://issues.apache.org/jira/browse/FLINK-24419 - not trimmed to 2
-                        // bytes
-                        .fromCase(STRING(), "Apache", new byte[] {65, 112, 97, 99, 104, 101})
+                        .fromCase(CHAR(3), "foo", new byte[] {102, 111})
+                        .fromCase(VARCHAR(5), "Flink", new byte[] {70, 108})
```

Comment:
Can you add a test for the non trim case where input length < target length? Both for the "type case" and for the "runtime case"

Reply:
Thanks for your detailed review!
I added more test casts to cover "type case" and "runtime case" when [converting to `VARBINARY`](https://github.com/apache/flink/pull/17919/files#diff-27c48372e941145405a73a03dde6ce99b5e926f2d4acdc8bb8cb9021a1486ec7R327), however, when converting to `BINARY(3)`, seems [Calcite 1.26 will pad the result](https://github.com/apache/calcite/blob/calcite-1.26.0/core/src/main/java/org/apache/calcite/rex/RexBuilder.java#L1539) to make sure it has the same length as target type, so converting `CHAR(1)` with value `f` to `BINARY(3)` will still result in `{102, 0, 0}`.

Reply:
@shenzhu Yep, this is something we have to implement as well. I'll open a separate issue for it.

Reply:
So, in case of this `CastFunctionITCase`, if we have padding it's totally fine, as calcite performs optimizations out of our control.

But i'm pretty sure that test case should work as described by me in `CastRulesTest`, without padding